compiler/cli/src/org/jetbrains/kotlin/cli/jvm/plugins/DependencyShadingPluginClassLoader.kt

```diff
@@ -0,0 +1,66 @@
+package org.jetbrains.kotlin.cli.jvm.plugins
+
+import org.jetbrains.org.objectweb.asm.ClassReader
+import org.jetbrains.org.objectweb.asm.ClassWriter
+import org.jetbrains.org.objectweb.asm.commons.ClassRemapper
+import org.jetbrains.org.objectweb.asm.commons.Remapper
+import java.io.IOException
+import java.io.InputStream
+import java.net.URL
+import java.net.URLClassLoader
+import java.security.AccessControlContext
+import java.security.AccessController
+import java.security.PrivilegedActionException
+import java.security.PrivilegedExceptionAction
+
+/**
+ * DependencyShadingPluginClassLoader will load a plugin jar, applying the necessary shading as required
+ * by the variant of the Kotlin Compiler which is consuming the plugin.
+ */
+class DependencyShadingPluginClassLoader(urls: Array<URL>, parent: ClassLoader) : URLClassLoader(urls, parent) {
+
+    val accessControlContext: AccessControlContext
+    val bytecodeLoader = URLClassLoader(urls, null)
+
+    init {
+        accessControlContext = AccessController.getContext()
+    }
+
+    // True if the Kotlin Compiler loading this plugin is shaded, and thus the plugin jar should also be shaded
+    val shouldBeShaded = com.intellij.mock.MockProject::class.java.name.startsWith("org.jetbrains.kotlin.")
+
+    @Throws(ClassNotFoundException::class)
+    override fun findClass(name: String): Class<*>? {
+        return try {
+            AccessController.doPrivileged(
+                PrivilegedExceptionAction {
+                    val path = name.replace('.', '/') + ".class"
+                    val originalBytes = bytecodeLoader.getResourceAsStream(path).readBytes()
+
+                    try {
+                        val cw = ClassWriter(ClassWriter.COMPUTE_MAXS)
+
+                        val shader = ClassRemapper(cw, object : Remapper() {
+                            override fun map(internalName: String): String {
+                                if(shouldBeShaded && internalName.startsWith("com/intellij/"))
+                                    return "org/jetbrains/kotlin/"+internalName
+                                if(!shouldBeShaded && internalName.startsWith("org/jetbrains/kotlin/com/intellij/"))
```

Comment:
I agree.  Long term, I would very much like the rule to be that nobody distributes shaded plugins.  However, it would be a breaking change for any existing 3rd party plugins.  I presumed that we would want to avoid such breaking changes.

That said, the plugin API has always been fairly unstable anyway, and there are only a small number of existing 3rd party plugins. If we are comfortable with this being a breaking change, I'm happy to make this change.

Would you like me to update this change us to enforce the new convention, or leave the CL as-is to avoid breaking changes?

Reply:
I'd say the main reason is the performance, I think it is quite expensive to run the ASM remapper in all cases, so to me, it is a good reason to break the compatibility. And I think that the real impact should be rather small, people who managed to make and distribute correctly shaded plugins may continue to use it in the shaded environment.

Reply:
JVM classloading is already slow, and ASM remapper is surprisingly fast.  But yes, I'm happy to switch this to skip the remapper if shading is off.  I'll make that change now.